http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/request/CrtRequestInputStreamAdapter.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.crt.internal.request;
+
+import static java.lang.Math.min;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.http.HttpRequestBodyStream;
+import software.amazon.awssdk.http.ContentStreamProvider;
+
+@SdkInternalApi
+final class CrtRequestInputStreamAdapter implements HttpRequestBodyStream {
+    private static final int READ_BUFFER_SIZE = 16 * 1024;
+
+    private final ContentStreamProvider provider;
+    private InputStream providerStream;
```

Comment:
I can't recall, does this need to be volatile? Or does CRT guarantee that the same thread always interacts with this?

Reply:
CRT doesn't guarantee that the same thread will always interact with this, but it does guarantee that the calls will be sequential and never overlapping, which removes the need for volatile.

Reply:
Just for context, I'm less worried about calls to overlapping, but ensuring that writes to `providerStream` by one thread, are visible by other threads that interact with this adapter.

Reply:
Yeah wanted to add for broader c interop comment. Volatile in Java is the same as an atomic load and store cpu instruction wise. So the case it helps is the visibility across threads (which have to actually be running already at the moment the write occurs for it to happen) if the read thread has a cached copy of the pointer. I dont actually think a modern chip needs it any more but anyways. 

If the consuming thread has a chance of doing a read of the values prior to it being written by another, you need volatile or you could get the cached version in theory though I’m not so sure it’s possible in practice anymore.

we should make it volatile just to be safe. Alternatively you can do the write inside the crt thread.